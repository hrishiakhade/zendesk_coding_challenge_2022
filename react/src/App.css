.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.ticket-item-description{
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  padding-right: 15px;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}



/* Modal */
.modal {
  background: white;
  border: .5px dotted lightgray;
  border-radius: 10px;
  width: 300px;
  outline: none;
  padding: 10px;
  text-align: center;
  margin-top: -200px;
  box-shadow: 0px 3px 15px rgba(207, 207, 207, 0.2);
}
.ReactModalPortal > div {
  opacity: 0;
  z-index: 99
}

.ReactModalPortal .ReactModal__Overlay {
  display: flex;
  justify-content: center;
  align-items: center;
  transition: opacity 200ms ease-in-out;
}

.ReactModalPortal .ReactModal__Overlay--after-open {
  opacity: 1;
}

.ReactModalPortal .ReactModal__Overlay--before-close {
  opacity: 0;
}

.button {
  background: #FCCDBF;
  border: none;
  padding: 10px;
  font-weight: 400;
  font-size: 1.1rem;
  border-radius: 10px;
  margin: 0 auto;
  width: 200px;
}

.dotted-line {
  border-bottom: .5px solid lightgray;
  padding-bottom: 10px;
}

.ticket-modal__description {
  padding: 15px;
  text-align: left;
  margin: 10px;
  border-bottom: 0.5px dotted lightgray;
}


.loader {
  margin: 200px auto;
  height: 20%;
  width: 20%;
}


.ticket-numbers-ul {
  max-width: 500px;
  list-style: none;
  display: -ms-flexbox;
  display: flex;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 1rem auto;
  overflow: hidden;
  padding: 5px;
}


.ticket-number-li{
  border: 0.5px solid lightgrey;
  padding: 10px;
  font-family: 'Montserrat', sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.error__message {
  color: red;
}
.error {
  background: white;
  border-radius: 5px;
  padding: 5px;
  margin: 40px auto;
}