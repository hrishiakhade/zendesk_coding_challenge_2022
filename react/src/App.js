import React, { Component, Touch } from "react";
import { getTickets } from "./utils/api-helpers";
import "./App.css";
import Header from "./components/Header";
import strings from "./utils/constants/strings.js";
import { getTotalPages, getPaginatedTickets } from "./utils/common-helpers";
import Pagination from "./components/Pagination";
import TicketList from "./components/TicketList";
import TicketDetailsModal from "./components/TicketDetailsModal";
import ErrorComponenet from "./components/ErrorMsg";
import loader from "./assets/loader.svg";

export default class App extends Component {
  state = {
    tickets: null,
    error: null,
    currentPage: 0,
    ticketsPerPage: 25,
    selectedTicket: null,
    totalTickets: 0,
    totalPages: 0,
    currentPageTickets: null,
    loading: false,
  };

  onClickTicket = (ticketData) => {
    const body = document.querySelector("body");
    //body.style.overflow = "hidden";
    this.setState({ selectedTicket: ticketData });
  };

  componentDidMount() {
    this.setState({ loading: true });
    getTickets()
      .then((res) => {
        if (res === 401) {
          this.setState({
            error: { message: strings.network_auth_error },
            loading: false,
          });
        } else if (res === 404) {
          this.setState({
            error: { message: strings.network_api_error },
            loading: false,
          });
        } else if (res === strings.network_error_header) {
          this.setState({
            error: { message: strings.network_api_error_msg },
            loading: false,
          });
        } else {
          this.state.totalTickets = res?.length ? res.length : 0;
          this.state.totalPages = getTotalPages(
            this.state.ticketsPerPage,
            this.state.totalTickets
          );
          this.state.tickets = getPaginatedTickets(
            this.state.ticketsPerPage,
            res,
            this.state.totalPages
          );
          this.setState({
            tickets: this.state.tickets,
            loading: false,
            currentPageTickets: this.state.tickets[this.state.currentPage],
            error: null,
          });
        }
      })
      .catch((error) => this.setState({ error, loading: false }));
  }

  onPageChange = (newPageNumber) => {
    if (newPageNumber > -1) {
      this.setState({
        currentPage: newPageNumber,
        currentPageTickets: this.state.tickets?.length
          ? this.state.tickets[newPageNumber]
          : [],
      });
    }
  };

  resetSelectedTicket = () => {
    this.setState({
      selectedTicket: null,
    });
    //const body = document.querySelector("body");
    //body.style.overflow = "auto";
  };

  render() {
    const {
      currentPage,
      totalPages,
      totalTickets,
      currentPageTickets,
      selectedTicket,
      error,
      loading,
    } = this.state;
    return (
      <div className="App container pt-3 pb-3" id="App">
        <Header />
        {loading ? (
          <img className="loader app-center" src={loader} alt="loading" />
        ) : error ? (
          <ErrorComponenet message={this.state?.error?.message} />
        ) : totalTickets > 0 ? (
          <div>
            <p>
              {strings.total_tickets} {totalTickets}
            </p>
            <TicketList
              currentPageTickets={currentPageTickets}
              onClickTicket={this.onClickTicket}
            />
            <br />
            <Pagination
              currentPage={currentPage}
              totalPages={totalPages}
              totalTickets={totalTickets}
              onPageChange={this.onPageChange}
            />

            <TicketDetailsModal
              selectedTicket={selectedTicket}
              resetSelectedTicket={this.resetSelectedTicket}
            />
          </div>
        ) : (
          <div>
            <p>{strings.no_tickets}</p>
          </div>
        )}
      </div>
    );
  }
}
