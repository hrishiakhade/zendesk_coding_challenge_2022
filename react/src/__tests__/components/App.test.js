import React from 'react';
import Enzyme, { shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json'
import App from '../../App';

Enzyme.configure({
  adapter: new Adapter()
})

test('Should render App correctly', () => {
  const wrapper = shallow(<App />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})
