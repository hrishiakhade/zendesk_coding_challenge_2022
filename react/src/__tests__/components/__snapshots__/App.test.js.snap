// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render App correctly 1`] = `
<div
  className="App container pt-3 pb-3"
  id="App"
>
  <Header />
  <img
    alt="loading"
    className="loader app-center"
    src="loader.svg"
  />
</div>
`;
